### ISSUE TITLE
Would it be possible to generate fonts without using the autoroute Satin?

### ISSUE THREAD

**Issue by karnigen** [inkstitch/inkstitch#2627]
Would it be possible to generate fonts without using the autoroute? Most or almost all fonts are generated with autoroute Satin.
Very often, the start and end points of stitches in characters are not in the ideal positions, and it is very laborious to fix it, especially when there are automatically generated objects by autoroute Satin.

It seems that it is much better to first place individual characters and then invoke autoroute only on the overall resulting text. Additionally, this allows for later adjustments of used font to accommodate national characters or combine with other objects.

Working with such fonts without autoroute would be:

- Place text with Inkstitch Lettering.
- Refine the positions of characters and/or add national symbols to the characters or whatever else.
- Use the autoroute Satin for stitching for overall result.

Ultimately, I'm not even clear on why characters in fonts are generated using autoroute. What advantage does it bring? I mostly see issues with it. It cannot be assumed that the next character would ideally begin where the previous one ends.

**Comment by Marif91** [comment]
Bonjour Karnigen,

J'ai pris trois polices de la liste, toutes en autoroute satin. Il y a très peu de sauts de fils, sauf entre les lignes, ils sont tous placés au même niveau.
Je ne vois pas où est le souci...
<img width="1691" alt="Capture d’écran 2023-12-08 à 16 17 01" src="https://github.com/inkstitch/inkstitch/assets/65076100/ef5b6e47-0a54-4bcc-ba9a-4883ef84cab7">

**Comment by karnigen** [comment]
It looks like a misunderstanding. Even in your example, some jumps are completely wrong, just take the pair of letters 'r' and 'n'.
The second line is generated as I suggested before. 

Or it's possible that your embroidery machine cuts the thread after the letter 'r' and continues to 'n'; then, of course, it's fine.
But my machine leaves the thread there even if I add a trim. It cuts the thread only when there's a color change, and I certainly don't want unnecessary jumps if they're not needed.


![2023-12-08_19-11](https://github.com/inkstitch/inkstitch/assets/7215249/6a12078a-8db2-468e-a2c6-5a86629ceb76)

**Comment by claudinepeyrat06** [comment]
What you are asking  implies that a letter would not always finish at the same place but that it would depend on the next letter. Therefore the preserve order option of autosatin could not be activated. But this is for the time being the only way to insure that some satin columns are stitched before some other one and this is often a requirement when stitching letters.
So  what you ask could probably be done by modifying the software , but not as simply as calling autosatin later on in the process.

**Comment by karnigen** [comment]
That is a very valid point. At the beginning, I didn't realize that.

On the other hand, some modification to the auto-satin algorithm might be good because sometimes it makes unnecessary jumps. I don't just mean characters, but also in other situations where a better solution clearly exists.

![2023-12-09_23-04](https://github.com/inkstitch/inkstitch/assets/7215249/ecee5e67-c733-4b63-8a96-5d2f8d446e60)
 

I can imagine an algorithm that would respect the order of satin columns in areas where the columns overlap and still end up at the desired end point. I can also envision overlapping satin columns where such a solution may not be possible, and a jump would be necessary.

Even in its current state, the program works excellently. Of course, the jumps are inconvenient because they require more time for embroidery cleanup.

**Comment by Augusa** [comment]
Je suis d'accord qu'une optimisation des sauts de fil serait bien mais ce n'est semble-t-il pas d'actualité. 
D'autre part je pense que la majorité des usagers veulent avoir du texte prêt à l'emploi et n'ont pas envie de manipuler des fonctionnalités complexes.
Dernier point, pour avoir créé une bonne quinzaine de polices d'Inkstitch, modifier un caractère pour lui ajouter un signe diacritique est simple et n'exige pas du tout de tout refaire! Heureusement!
Ce que beaucoup d'usagers souhaiterait sans aucun doute c'est la possibilité de prendre une police TTF de leur choix et de pouvoir la transformer facilement en broderie. Ça manque réellement dans Inkstitch

**Comment by kaalleen** [comment]
If you know how to improve the auto-satin algorithm we'd be really happy to take in pull requests :)

**Comment by karnigen** [comment]
Really not sure if I can handle it. I can't even debug that.

So that I could even start with something, I have created a fork and added support for VSCode editor debugging and offline debugging. Now DEBUG file may contain the name of required editor debugger support: _vscode, pycharm, pydev, none_
or _vscode-script, pycharm-script, pydev-script_. Also file may contain comments or empty lines.

In the second case, _-script_ variant,
- when called from inkscape then local script for offline debugging `.ink.sh` and `.ink.svg` is created and script is running without debug
- when you run script `.ink.sh` from console, debug is enabled, or may be disabled by DEBUG file.

I've committed changes in my kardev branch in my fork.

But what's next? I'm not an expert on graphical interfaces. I need a general idea of where to place a breakpoint in wxWindow when custom algorithms are called. Additionally, how to debug Flask, as the debugger doesn't seem to activate in the browser for me. It seems that many outputs are intentionally suppressed for the release version. How can I enable them for development? 
Is there any documentation for developers on how to debug and work with it?

**Comment by kaalleen** [comment]
https://inkstitch.org/developers/inkstitch/debug/

Does it help you to get started?

**Comment by karnigen** [comment]
Thank you for the guidance. It seems that the information on web is concise enough. 

Now it seem that debug and profile is working but only after refactoring `inkstitch.py`, extending remote debug for **vscode** editor and direct starting from debugger, extending profiling with `pyinstrument` and simpler `profile` modules next to `cProfile`.
Type of editor is written in `DEBUG` file, and type of profiler module in `PROFILE` file. To disable debug or profile simply comment them with `#` in these files, not needed to delete, rename or move files. To change editor or profile module comment one and uncomment another in `DEBUG` or `PROFILE` files.
Also added possibility to simulate call _inkstitch_ from _inkscape_ without actually run _inkscape_ (offline debug). 

Additionally, I have added many comments to the `inkstitch.py` so that beginners can orient themselves more quickly.
I hope it will be useful to someone, see my fork. I hope I haven't changed anything, and it's backward compatible; developers just need to add their editor (pydev for LiClipse, pycharm for PyCharm, vscode for Visual Studio Code) and profiler (cprofile, pyinstrument, profile) to the specified files.

Now I need to know idea how svg is parsed and how stitches are generated, if possible, without using graphics. Or what is call hierarchy.

**Comment by kaalleen** [comment]
I did send you an invitation to join the inkstitch community as a developer. This way you can create branches in this repository and use the build system, so everyone will be able to testrun your builds.

Here a bunch of links which might we useful for the task:

https://github.com/inkstitch/inkstitch/issues/1913
https://inkstitch.org/developers/inkstitch/coding-style/
https://www.youtube.com/live/nzAJKkX4N1s?t=1475

**Comment by karnigen** [comment]
I created a branch in kgn/debug_profile_extend_vscode that I filled with changes as mentioned above. 
How do I perform tests for other systems or what next? Who can accept or reject the changes?
Will the developers agree with this solution?

I've tested it on Linux with VS Code editor and all profilers. Other editors (PyCharm and LiClipse) I'm not currently using.

**Comment by kaalleen** [comment]
When you make it a pull request you can ping @lexelby and ask for a review.

**Comment by bkmgit** [comment]
The fonts portion would need a re-write and perhaps follow a similar strategy to how fonts are rendered.  This would perhaps allow automatic combining of diacritics and placement of ligatures used in Arabic, Indic and similar scripts.  For Chinese and Japanese fonts, unclear what is best.

**Comment by kaalleen** [comment]
I think this comes down to a request of a better autoroute algorithm for satins. We already have this request open.
So I'll close this issue here. Please protest, if I misunderstood.
